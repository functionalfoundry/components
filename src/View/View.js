/* @flow */
import React from 'react'
import Theme from 'js-theme'
import applyLayout from './modules/applyLayout'
import applyNativeMethods from './modules/applyNativeMethods'
import createDOMElement from './modules/createDOMElement'
import normalizeNativeEvent from './modules/normalizeNativeEvent'

type PointerEventsT = 'auto' | 'box-none' | 'box-only' | 'none'
type AccessibilityLiveRegionT = 'assertive' | 'off' | 'polite'
type EdgeInsetsT = {
  top: number,
   left: number,
   bottom: number,
   right: number,
}

type PropsT = {
  children?: React.Children,
  inline?: boolean,
  testID?: string,
  theme: Object,
  accessibilityLabel: string,
  accessibilityLiveRegion: AccessibilityLiveRegionT,
  accessibilityRole: string,
  accessible: bool,
  collapsable: bool,
  hitSlop: EdgeInsetsT,
  onClick: Function,
  onClickCapture: Function,
  onLayout: Function,
  onMoveShouldSetResponder: Function,
  onMoveShouldSetResponderCapture: Function,
  onResponderGrant: Function,
  onResponderMove: Function,
  onResponderReject: Function,
  onResponderRelease: Function,
  onResponderTerminate: Function,
  onResponderTerminationRequest: Function,
  onStartShouldSetResponder: Function,
  onStartShouldSetResponderCapture: Function,
  onTouchCancel: Function,
  onTouchCancelCapture: Function,
  onTouchEnd: Function,
  onTouchEndCapture: Function,
  onTouchMove: Function,
  onTouchMoveCapture: Function,
  onTouchStart: Function,
  onTouchStartCapture: Function,
  pointerEvents: PointerEventsT,
}

const defaultProps = {
  accessible: true,
  theme: {},
}

const View = ({
  inline,
  theme,
  collapsable, // eslint-disable-line
  hitSlop, // eslint-disable-line
  onLayout, // eslint-disable-line
  pointerEvents, // eslint-disable-line
  ...props,
}: PropsT) => {
  const normalizedEventHandlers = eventHandlerNames.reduce((handlerProps, handlerName) => {
    const handler = props[handlerName]
    if (typeof handler === 'function') {
      handlerProps[handlerName] = normalizeEventForHandler(handler, handlerName)
    }
    return handlerProps
  }, {})

  const component = inline ? 'span' : 'div'
  const finalProps = {
    ...props,
    ...theme.view,
    ...normalizedEventHandlers,
  }

  return createDOMElement(component, finalProps)
}

View.defaultProps = defaultProps
View.displayName = 'View'

const defaultTheme = ({
  inline,
  pointerEvents,
}: PropsT) => ({
  view: {
    borderWidth: 0,
    borderStyle: 'solid',
    boxSizing: 'border-box',
    margin: 0,
    padding: 0,
    position: 'relative',
    // button and anchor reset
    backgroundColor: 'transparent',
    color: 'inherit',
    font: 'inherit',
    textAlign: 'inherit',
    textDecorationLine: 'none',
    // list reset
    listStyle: 'none',
    // fix flexbox bugs
    minHeight: 0,
    minWidth: 0,
    ...getFlexStyle(inline || false), // TODO: Change to default?
    ...getPointerEventsStyle(pointerEvents),
  },
})

const getFlexStyle = (inline: boolean) => {
  if (inline) {
    return {}
  }
  return {
    alignItems: 'stretch',
    display: 'flex',
    flex: '1 0 auto',
    flexDirection: 'column',
  }
}

const getPointerEventsStyle = (pointerEvents: PointerEventsT) => {
  if (pointerEvents) {
    return { pointerEvents }
  }
  return {}
}

const eventHandlerNames = [
  'onClick',
  'onClickCapture',
  'onMoveShouldSetResponder',
  'onMoveShouldSetResponderCapture',
  'onResponderGrant',
  'onResponderMove',
  'onResponderReject',
  'onResponderRelease',
  'onResponderTerminate',
  'onResponderTerminationRequest',
  'onStartShouldSetResponder',
  'onStartShouldSetResponderCapture',
  'onTouchCancel',
  'onTouchCancelCapture',
  'onTouchEnd',
  'onTouchEndCapture',
  'onTouchMove',
  'onTouchMoveCapture',
  'onTouchStart',
  'onTouchStartCapture',
]

const normalizeEventForHandler = (handler, handlerName) => {
  // Browsers fire mouse events after touch events. This causes the
  // ResponderEvents and their handlers to fire twice for Touchables.
  // Auto-fix this issue by calling 'preventDefault' to cancel the mouse
  // events.
  const shouldCancelEvent = handlerName.indexOf('onResponder') === 0

  return (e) => {
    e.nativeEvent = normalizeNativeEvent(e.nativeEvent)
    const returnValue = handler(e)
    if (shouldCancelEvent && e.cancelable) {
      e.preventDefault()
    }
    return returnValue
  }
}

export default Theme('View', defaultTheme)(applyLayout(applyNativeMethods(View)))
