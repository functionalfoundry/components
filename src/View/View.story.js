import React from 'react'
import { storiesOf } from '@kadira/storybook'
import View from './View'
import PreviewContainer from '../PreviewContainer/PreviewContainer'
import Preview from '../Preview'

storiesOf('View', module)
  .add('Regular', () => (
    <PreviewContainer
      shade='dark'
    >
      <Preview
        label='Without props'
      >
        <View>
          {'View'}
        </View>
      </Preview>
      <Preview
        label='With theme'
      >
        <View
          theme={getTheme()}
        >
          {'Content text'}
        </View>
      </Preview>
      <Preview
        label='Inline'
      >
        <View
          theme={getTheme()}
          inline
        >
          {'Content text'}
        </View>
      </Preview>
      <Preview
        label='Without props'
      >
        <View>
          {'View'}
        </View>
      </Preview>
      <Preview
        label='Nested props'
      >
        <View
          theme={getTheme()}
        >
          {'1'}
        </View>
        <View
          theme={getTheme('green')}
        >
          {'2'}
        </View>
        <View>
          <View
            theme={getTheme('yellow')}
          >
            {'3'}
          </View>
          <View
            theme={getTheme('red')}
          >
            {'4'}
          </View>
        </View>
      </Preview>
      <Preview
        label='With test ID'
      >
        <View
          testId='View234'
        >
          {'View'}
        </View>
      </Preview>
    </PreviewContainer>
  ))

const getTheme = (color = 'aqua') => ({
  view: {
    backgroundColor: color,
  },
})
