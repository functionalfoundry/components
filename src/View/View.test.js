import React from 'react'
import View from '.'

test('View renders', () => {
  expect(<View />).toMatchSnapshot()
})

test('View renders theme', () => {
  expect(
    <View
      theme={{ view: {} }}
    />
  ).toMatchSnapshot()
})

test('View renders inline state', () => {
  expect(
    <View
      inline
    />
  ).toMatchSnapshot()
})
