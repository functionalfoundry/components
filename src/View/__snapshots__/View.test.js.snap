exports[`test View renders 1`] = `
<function Themed(props) {
        _classCallCheck(this, Themed);

        var _this = _possibleConstructorReturn(this, (Themed.__proto__ || Object.getPrototypeOf(Themed)).call(this, props));

        _this.state = {
          theme: {},
          styleSheet: {}
        };
        _this.computedLocalTheme = undefined;
        return _this;
      }
  composeTheme="deeply" />
`;

exports[`test View renders inline state 1`] = `
<function Themed(props) {
        _classCallCheck(this, Themed);

        var _this = _possibleConstructorReturn(this, (Themed.__proto__ || Object.getPrototypeOf(Themed)).call(this, props));

        _this.state = {
          theme: {},
          styleSheet: {}
        };
        _this.computedLocalTheme = undefined;
        return _this;
      }
  composeTheme="deeply"
  inline={true} />
`;

exports[`test View renders theme 1`] = `
<function Themed(props) {
        _classCallCheck(this, Themed);

        var _this = _possibleConstructorReturn(this, (Themed.__proto__ || Object.getPrototypeOf(Themed)).call(this, props));

        _this.state = {
          theme: {},
          styleSheet: {}
        };
        _this.computedLocalTheme = undefined;
        return _this;
      }
  composeTheme="deeply"
  theme={
    Object {
      "view": Object {}
    }
  } />
`;
